src/main/java/appland/telemetry/AppMapRecordListener.java

```diff
@@ -0,0 +1,32 @@
+package appland.telemetry;
+
+import org.jetbrains.annotations.NotNull;
+
+import com.intellij.openapi.Disposable;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.startup.StartupActivity;
+
+import appland.files.AppMapFileChangeListener;
+
+public class AppMapRecordListener implements Disposable, StartupActivity {
```

Comment:
It's better to register this as a listener inside `<applicationListeners>` inside `plugin.xml`:
```xml
    <applicationListeners>
        ...
        <listener topic="appland.files.AppMapFileChangeListener"
                  class="appland.telemetry.AppMapRecordListener"/>
    </applicationListeners>

```

with this implementation:

```java
package appland.telemetry;

import appland.files.AppMapFileChangeListener;
import org.jetbrains.annotations.NotNull;

import java.util.concurrent.atomic.AtomicBoolean;

public class AppMapRecordListener implements AppMapFileChangeListener {
    private final @NotNull AtomicBoolean appMapRecordedThisSession = new AtomicBoolean(false);

    @Override
    public void afterAppMapFileChange() {
        if (appMapRecordedThisSession.compareAndSet(false, true)) {
            TelemetryService.getInstance().sendEvent("appmap:record");
        }
    }
}
```

`AtomicBoolean` (or a `volatile boolean`) is preferred over a `Boolean` to avoid potential threading issues, e.g. by concurrent updates of the field.

Please note that it's possible that `afterAppMapFileChange` is called after a deletion of an `.appmap.json` file.

Reply:
Thanks! I've switched the startup activity to a listener.

> Please note that it's possible that `afterAppMapFileChange` is called after a deletion of an `.appmap.json` file.

In regard to this, I had to change the  `AppMapAsyncFileListener` interface a bit to inform the listener about the types of changes that have occurred. This felt a bit better than creating a new type of listener that notifies only on create or update. Let me know what you think.